### ISSUE TITLE
Spurious error when copying a list via slice

### ISSUE THREAD

**Issue by akaptur** [python/mypy#5335]
Mypy correctly emitted an error on some code like this:
``` python
from typing import List, Union

def f(a: List[Union[str, int]]) -> None:
    return None

x: List[str] = []

f(x)  # error: Argument 1 to "f" has incompatible type "List[str]"; expected "List[Union[str, int]]"
```
My caller won't be modifying `x` and doesn't need to share a reference to it, so I copied the list using a comprehension, which also satisfied mypy. 

I expect any method of copying the list to satisfy mypy, but in practice only two of the three common ways to copy a list work. Mypy continues to emit the error when I copy the list using a slice.
``` python
f(list(x))  # ok
f([elem for elem in x])  # ok
f(x[:])  # error: Argument 1 to "f" has incompatible type "List[str]"; expected "List[Union[str, int]]"
```

I expect the third form to behave the same way as the first two. 

I'm on mypy 0.610, Python 3.6.5.

**Comment by gvanrossum** [comment]
Hi Allison, the recommended solution in this case is to declare `f()` as taking a `Mapping` rather than a `List`.

**Comment by ilevkivskyi** [comment]
> I expect the third form to behave the same way as the first two.

There are some subtle reasons why it is impossible without reducing type safety, for example a user subclass can override `__getitem__` to return a view instead of a copy (although this is of course a very rare scenario). I think instead of adding some special casing in mypy you can just use an immutable container as suggested (you can also use `Sequence`).
